pydoctor/astbuilder.py

```diff
@@ -512,36 +485,81 @@ def _annotation_from_attrib(self, expr, ctx):
+        @param func: The function definition's AST.
+        @return: Mapping from argument name to annotation.
+            The name C{return} is used for the return type.
+            Unannotated arguments are omitted.
+        """
+        def _get_all_args() -> Iterator[ast.arg]:
+            base_args = func.args
+            # New on Python 3.8 -- handle absence gracefully
+            try:
+                yield from base_args.posonlyargs
+            except AttributeError:
+                pass
+            yield from base_args.args
+            varargs = base_args.vararg
+            if varargs:
+                yield varargs
+            yield from base_args.kwonlyargs
+            kwargs = base_args.kwarg
+            if kwargs:
+                yield kwargs
+        def _get_all_ast_annotations() -> Iterator[Tuple[str, ast.expr]]:
+            for arg in _get_all_args():
+                if arg.annotation:
+                    yield arg.arg, arg.annotation
+            returns = func.returns
+            if returns:
+                yield 'return', returns
+        return {name: self._unstring_annotation(value)
+                for name, value in _get_all_ast_annotations()}
+
+    def _unstring_annotation(self, node: ast.expr) -> ast.expr:
         """Replace all strings in the given expression by parsed versions.
-        Returns the resulting node, or None if parsing failed.
+        @return: The unstringed node. If parsing fails, an error is logged
+            and the original node is returned.
         """
         try:
-            return _AnnotationStringParser().visit(node)
+            expr = _AnnotationStringParser().visit(node)
+            assert isinstance(expr, ast.expr), expr
+            return expr
         except SyntaxError as ex:
             self.builder.currentMod.report(
                     f'syntax error in annotation: {ex}',
                     lineno_offset=node.lineno)
-            return None
+            return node
 
 
 class _AnnotationStringParser(ast.NodeTransformer):
+    """Implementation of L{ModuleVistor._unstring_annotation()}."""
+
+    def _parse_string(self, value: str) -> ast.expr:
+        stmt, = ast.parse(value).body
+        if isinstance(stmt, ast.Expr):
+            expr = self.visit(stmt.value)
+            assert isinstance(expr, ast.expr), expr
+            return expr
+        else:
+            raise SyntaxError("expected expression, found statement")
 
     # For Python >= 3.8:
 
-    def visit_Constant(self, node):
+    def visit_Constant(self, node: ast.Constant) -> ast.expr:
         value = node.value
         if isinstance(value, str):
-            expr, = ast.parse(value).body
-            return self.visit(expr.value)
+            return self._parse_string(value)
         else:
-            return self.generic_visit(node)
+            const = self.generic_visit(node)
+            assert isinstance(const, ast.Constant), const
```

Comment:
Why is this assert here? A potential bug in `generic_visit`?

Reply:
This one is mostly for mypy. If `generic_visit()` changes the type of node, that would indeed be a bug in `NodeTransformer`, I think. So I could change this into a cast, but I still prefer an assert for more certainty.